service/history/api/respondworkflowtaskcompleted/api.go

```diff
@@ -162,17 +167,48 @@ func (handler *WorkflowTaskCompletedHandler) Invoke(
 	}
 	weContext := workflowLease.GetContext()
 	ms := workflowLease.GetMutableState()
-
 	currentWorkflowTask := ms.GetWorkflowTaskByID(token.GetScheduledEventId())
+	defer func() {
+		if retError != nil && currentWorkflowTask != nil && currentWorkflowTask.Type == enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE && ms.IsStickyTaskQueueSet() {
+			// If, while completing WFT, error is occurred and returned to the worker then worker will clear its cache.
+			// New WFT will also be created on sticky task queue and sent to worker.
+			// Because worker doesn't have a workflow in cache, it needs to replay it from the beginning,
+			// but sticky WFT has only partial history. Worker will request full history using GetWorkflowExecutionHistory API.
+			// This history doesn't have speculative WFT events. If WFT is speculative,
+			// then worker will see inconsistency between history from it and full history, and will fail WFT.
+			//
+			// To prevent unexpected WFT failure, server clears stickiness for this workflow, next WFT will go to normal task queue,
+			// and will have full history attached to it.
+			// This is NOT 100% bulletproof solution because this write operation may also fail.
+			// TODO: remove this call when GetWorkflowExecutionHistory includes speculative WFT events.
+			if clearStickyErr := handler.clearStickyTaskQueue(ctx, workflowLease.GetContext()); clearStickyErr != nil {
```

Comment:
now we should release lock with the clearStickyErr?

Reply:
In practice it doesn't matter, because in case of error WF context is cleared no matter what is the error type. In future, though, I plan to exclude some error types (like query errors) and not clear workflow context on them. And WF context definitely needs to be cleared after error here, so I am changing it to `clearStickyErr`.